AndroidXCI/lib/src/main/kotlin/dev/androidx/ci/testRunner/TestRunnerServiceImpl.kt

```diff
@@ -204,6 +181,61 @@ internal class TestRunnerServiceImpl internal constructor(
         }
     }
 
+    private suspend fun findLogcats(
+        testMatrix: TestMatrix
+    ): MutableMap<TestRunnerService.TestIdentifier, TestRunnerService.ResultFileResource> {
+        var testCaseLogcatBlobs = mutableMapOf<TestRunnerService.TestIdentifier, TestRunnerService.ResultFileResource>()
+        val resultPath = testMatrix.resultStorage.googleCloudStorage.gcsPath
+
+        // get cached result if available, else invoke tool results apis to get test execution steps and store them in glcoud bucket
+        val logcatsFilePath = "$resultPath/logcats"
+        val logcatsRelativePath = logcatsFilePath.substringAfter("gs://").substringAfter("/")
+        if (googleCloudApi.existingFilePath(logcatsRelativePath) != null) {
+            googleCloudApi.walkEntries(GcsPath(logcatsFilePath)).first().obtainInputStream().reader().readLines().forEach {
+                val className = it.split("#")[0]
+                val name = it.split("#")[1]
+                val runNumber = it.split("#")[2]
+                val path = it.split("#")[3]
+                testCaseLogcatBlobs[
+                    TestRunnerService.TestIdentifier(
+                        className,
+                        name,
+                        runNumber.toInt()
+                    )
+                ] = ResultFileResourceImpl(googleCloudApi.walkEntries(GcsPath(path)).first())
+            }
+        } else {
+            fun BlobVisitor.fullDeviceId() = relativePath.substringBefore('/', "")
+            val steps = testExecutionStore.getTestExecutionSteps(testMatrix)
+            val out: java.lang.StringBuilder = StringBuilder()
+            googleCloudApi.walkEntries(
+                gcsPath = GcsPath(resultPath)
+            ).forEach { visitor ->
+                if (visitor.fileName.endsWith(LOGCAT_FILE_NAME_SUFFIX)) {
+                    val step = steps.flatMap {
+                        it.testExecutionStep?.toolExecution?.toolOutputs ?: emptyList()
+                    }?.find {
+                        (it.output?.fileUri == visitor.gcsPath.toString())
+                    }
+                    val runNumber = DeviceRun.create(visitor.fullDeviceId()).runNumber
+                    step?.testCase?.className?.let { className ->
+                        step?.testCase?.name?.let { name ->
+                            TestRunnerService.TestIdentifier(
+                                className,
+                                name,
+                                runNumber
+                            )
+                        }
+                    }?.let { testIdentifier ->
+                        testCaseLogcatBlobs[testIdentifier] = ResultFileResourceImpl(visitor)
+                        out.appendLine("$testIdentifier#${visitor.gcsPath.path}")
```

Comment:
we should use a proper serializable object here (likely json since we already use it elsewhere)

that would also affect how you read it above (basically stream from GCP and turn that into the object that you'll return).